### PR TITLE
Convolution shape inference

### PR DESCRIPTION
**Pull request by jane-intel**
### Details: 
 - Precise spatial shape inference for dynamic shapes
  - Deprecated set_adding_above method and given alternative set_pads_end
  - Enabled shape inference for default-constructed Convultions (private field `num_spatials` is no more required to call shape_infer)
  - Test adjusted for dynamic spatial shapes calculation
  - Style check enabled for shape_inference lib -- sorry for the mess

### Tickets:
 - *82949, 82963*

### CODE REVIEW

**Comment by jane-intel** [comment]
CI check for regressions is in progress right now. Waiting green light from @rnugmanx
Until then -- DO NOT MERGE, but review is necessary

**Comment by akladiev** [comment]
The valid link to the validation results: https://openvino-ci.toolbox.iotg.sclab.intel.com/job/private-ci/job/github_trigger/job/openvino/job/PR-12602/1/

File: src/core/include/openvino/op/convolution.hpp
Comment (ilyachur) [inline src/core/include/openvino/op/convolution.hpp:119]: Why do we need to introduce new function? It is a private API, I think we can extend the signature of legacy function instead of introducing new function and new friends.

File: src/core/include/openvino/op/convolution.hpp
Comment (ilyachur) [inline src/core/include/openvino/op/convolution.hpp:300]: The same comment

File: src/core/shape_inference/CMakeLists.txt
Comment (ilyachur) [inline src/core/shape_inference/CMakeLists.txt:25]: 👍

File: src/core/shape_inference/include/convolution_shape_inference.hpp
Comment (praasz) [inline src/core/shape_inference/include/convolution_shape_inference.hpp:546]: it looks  like this both case are excludes so use some `if..else` to reduce number of checks and assignments to `num_spatial`

Also this block code (L:545-560) looks same as L:625-642, make some helper to reduce duplicates?

File: src/core/shape_inference/include/convolution_shape_inference.hpp
Comment (praasz) [inline src/core/shape_inference/include/convolution_shape_inference.hpp:540]: Can be `constexpr`?

File: src/core/tests/type_prop/convolution.cpp
Comment (praasz) [inline src/core/tests/type_prop/convolution.cpp:146]: Missing test assert/except macro?

File: src/core/tests/type_prop/convolution_backprop_data.cpp
Comment (praasz) [inline src/core/tests/type_prop/convolution_backprop_data.cpp:839]: Missing test assert/except macro?
Check other similar added test it looks there is no `gtest` expectation macro.

File: src/plugins/intel_cpu/src/utils/shape_inference/static_dimension.hpp
Comment (praasz) [inline src/plugins/intel_cpu/src/utils/shape_inference/static_dimension.hpp:31]: Can you explain variable name `ldimension` in comment `lower bound of dimension`?
In `()` is mentioned that this variable must be equal to `ldimension`, itself. Is that ok? shoud `udimension`

Check also description of `udimension`

File: src/core/shape_inference/include/convolution_shape_inference.hpp
Comment (jane-intel) [reply src/core/shape_inference/include/convolution_shape_inference.hpp:561]: this method has varying number of parameters so this is not a duplication

File: src/core/shape_inference/include/convolution_shape_inference.hpp
Comment (jane-intel) [reply src/core/shape_inference/include/convolution_shape_inference.hpp:268]: this is a bit more explicit and compiler would do the optimization anyway

**Review by ilyachur — APPROVED** [review]

**Pull request merged**
